File: python/two-bucket/two_bucket.py
def measure(bucket_one, bucket_two, goal, start_bucket):
    if (goal > bucket_one and goal > bucket_two):
        raise ValueError("too big")

    possibles = list()
    if start_bucket == "one":
        _measure(bucket_one, bucket_one, bucket_two, 0, 1, goal, start_bucket, set(), possibles)
        if possibles: return min(possibles)
        else: raise ValueError("impossible")
    else:
        _measure(bucket_one, 0, bucket_two, bucket_two, 1, goal, start_bucket, set(), possibles)
        if possibles: return min(possibles)
        else: raise ValueError("impossible")

def _measure(b1_size, b1_amt, b2_size, b2_amt, actions, goal, start_bucket, seen, possibles):
    _tuple = (b1_size, b1_amt, b2_size, b2_amt)
    if _tuple in seen: return
    seen.add(_tuple)
    if b1_amt == goal:
        possibles.append((actions, "one", b2_amt))
        return
    if b2_amt == goal:
        possibles.append((actions, "two", b1_amt))
        return
    # empty b1
    if b1_amt > 0 and not (start_bucket == "one" and b2_amt == b2_size):
        if _measure(b1_size, 0, b2_size, b2_amt, actions + 1, goal, start_bucket, seen, possibles):
            return
    # empty b2
    if b2_amt > 0 and not (start_bucket == "two" and b1_amt == b1_size):
        if _measure(b1_size, b1_amt, b2_size, 0, actions + 1, goal, start_bucket, seen, possibles):
            return
    # fill b1
    if b1_amt < b1_size and not (start_bucket == "two" and b2_amt == 0):
        if _measure(b1_size, b1_size, b2_size, b2_amt, actions + 1, goal, start_bucket, seen, possibles):
            return
    # fill b2
    if b2_amt < b2_size and not (start_bucket == "one" and b1_amt == 0):
        if _measure(b1_size, b1_amt, b2_size, b2_size, actions + 1, goal, start_bucket, seen, possibles):
            return
    # pour b1 into b2
    if b1_amt > 0 and b2_amt < b2_size:
        _b1_amt = 0
        _b2_amt = b1_amt + b2_amt
        if _b2_amt > b2_size:
            _b1_amt = _b2_amt - b2_size
            _b2_amt = b2_size
        if not (start_bucket == "one" and _b2_amt == b2_size and _b1_amt == 0):
            if _measure(b1_size, _b1_amt, b2_size, _b2_amt, actions + 1, goal, start_bucket, seen, possibles):
                return

    # pour b2 into b1
    if b2_amt > 0 and b1_amt < b1_size:
        _b2_amt = 0
        _b1_amt = b2_amt + b1_amt
        if _b1_amt > b1_size:
            _b2_amt = _b1_amt - b1_size
            _b1_amt = b1_size
        if not (start_bucket == "two" and _b1_amt == b1_size and _b2_amt == 0):
            if _measure(b1_size, _b1_amt, b2_size, _b2_amt, actions + 1, goal, start_bucket, seen, possibles):
                return

    return

File: ruby/two-bucket/two_bucket.rb
class TwoBucket

  attr_reader :moves, :goal_bucket, :other_bucket

  def initialize(b1_size, b2_size, goal, start_bucket)
    sizes = [b1_size, b2_size]
    shortest = nil
    seen = {}

    choose = ->(amts, action, actions, active) {
      other = active == 1 ? 0 : 1
      case action
      when "fill"
        amts[active] = sizes[active]
      when "empty"
        amts[active] = 0
      when "pour"
        amt_to_pour = amts[other]
        space = sizes[active] - amts[active]
        amt_to_pour = space if amt_to_pour > space
        amts[active] += amt_to_pour
        amts[other] -= amt_to_pour
      end
      actions << amts

      if (!shortest || actions.size < shortest) and amts.include?(goal)
        @moves = actions.size
        @goal_bucket = amts[0] == goal ? "one" : "two"
        @other_bucket = (@goal_bucket == "one") ? amts[1] : amts[0]
        return
      end
      # avoid infinite loops
      return if seen[amts.join(",")]
      seen[amts.join(",")] = 1

      choose.call(amts.dup, "pour",  actions.dup, active) if amts[other] > 0 and amts[active] < sizes[active]
      choose.call(amts.dup, "pour",  actions.dup, other)  if amts[active] > 0 and amts[other] < sizes[other]
      choose.call(amts.dup, "fill",  actions.dup, active) if amts[active] < sizes[active]
      choose.call(amts.dup, "fill",  actions.dup, other)  if amts[other] < sizes[other]
      choose.call(amts.dup, "empty", actions.dup, active) if amts[active] > 0
      choose.call(amts.dup, "empty", actions.dup, other)  if amts[other] > 0
    }
    choose.call([0,0], "fill", [], start_bucket == "one" ? 0 : 1)
  end
end
